### ISSUE TITLE
Implement message logging functionality

### ISSUE THREAD

**Issue by bbatsov** [nrepl/nrepl#85]
**Is your feature request related to a problem? Please describe.**

Often when people encounter an issue with nREPL the best way to debug the issues is to expect the message exchange between their client and nREPL, so we/they can check whether the request and response messages are what we're expecting them to be (both in their encoded and decoded form). 

**Describe the solution you'd like**

Ideally we'll need some way to log all messages as Clojure maps and in their encoded format (e.g. bencode) to a file. Perhaps we can have a logging middleware that checks for some `:verbose` key in messages and would log those. Clients can easily implement a toggle turning on/off message logging.  We can probably have different options for showing the messages encoded or decoded. Likely we should have one log file per each server that goes to `.nrepl/` or the current directory, so logs for different servers don't get interleaved (or we can have one log file and just prefix messages accordingly there). 

**Describe alternatives you've considered**

CIDER has some client-side message logging functionality, but there's no easy way to inspect what happens with the message on the server's end. 

This doesn't really have to be a middleware - it can also be some server config (e.g. we can add a `verbose` flag to `start-server` and just check for it on the transport level), but I think a middleware will be a more flexible solution.  

Looking at #79 it seems that @pfeodrippe should be able to easy handle this. Given his work on #84 he should also immediately appreciate the importance of having this. :D

**Comment by pfeodrippe** [comment]
Definitively I'm going to tackle something about this at #84, thanks for this issue, it's gonna be a good guide! o/

**Comment by pfeodrippe** [comment]
I'm having some problems with it
1) When reading data, `transport-read` function is not available at `FnTransport`, can't see the data without changing the concrete transport implementation
2) `send-fn` at is opaque

For `1`, we could get the message after decoded and encode it again, but maybe it would not be representative enough of the received message.

For 2, I can compose the message using the `send-fn` field (if `FnTransport` becomes a Record) or use the reversed proposed solution for `1`

Maybe would be better to snitch the data for `in` and `out` streams, but using this would require a global configuration instead per message.

Any other options suggested?

**Comment by bbatsov** [comment]
> When reading data, transport-read function is not available at FnTransport, can't see the data without changing the concrete transport implementation

So using something like `logging-transport` wrapper won't work? I think this was my original idea, but I didn't think much about this. 

> Maybe would be better to snitch the data for in and out streams, but using this would require a global configuration instead per message.

Yeah, perhaps that'd be better. We can just add some `debug/verbose` config to `start-server` and potentially support this via `nrepl.config`. That won't allow us to toggle dynamically the logging on/off, but it'd be a perfectly adequate solution for me.

**Comment by cgrand** [comment]
There's something about the scope of this issue which is not clear to me: what kind of issues are we trying to debug? Transport/serialization issues or applicative issues (non-conforming messages etc.)?
I believe we shouldn't try to address both of them at once.

**Comment by bbatsov** [comment]
> Transport/serialization issues or applicative issues (non-conforming messages etc.)?

Both. :-)

> I believe we shouldn't try to address both of them at once.

Yeah, I agree with you, therefore my suggestion to split this PR in two and focus on the more common problem - the non-conforming messages. 

Transport/serialization issues weren't very common in the past (mostly because everyone was using the same transport, which was pretty well-tested), but I assume that when we start rolling out new transports there might be some rough edges there, that's why some mechanism to monitor the serialization would be useful.

**Comment by devurandom** [comment]
I am trying to debug an issue where internal messages sent from Cursive (i.e. not sent by the user) are printed in the Cursive REPL, instead of hidden. Colin Fleming suspects the response message might have a matching ID anymore, but to confirm that, we would need logs of the nREPL messages that are actually sent over the wire. Is there already a solution for this in nREPL?

**Comment by bbatsov** [comment]
@devurandom Nope, we never managed to sort out message logging on the server-side. There was some work done in https://github.com/nrepl/nrepl/pull/87, but it was not completed.

**Comment by devurandom** [comment]
@plexus just created https://github.com/lambdaisland/nrepl-proxy as a debugging tool for nREPL connections! :heart: :tada:

**Comment by plexus** [comment]
Yup, something I've wanted for a long time. I decided to do it as a proxy to prevent middleware ordering issues. This way you're sure you get the messages as they go over the wire. It's also nice that you can just pipe this to a file and share it with developers.

I have some ideas of where to take this, for now it's pretty basic. I mainly wanted to sort out the network part, the rest should be straightforward.

**Comment by bbatsov** [comment]
Cool idea!
